src/core/server/saved_objects/export/saved_objects_exporter.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Boom from '@hapi/boom';
+import { createListStream } from '@kbn/utils';
+import { PublicMethodsOf } from '@kbn/utility-types';
+import { SavedObject, SavedObjectsClientContract } from '../types';
+import { fetchNestedDependencies } from './fetch_nested_dependencies';
+import { sortObjects } from './sort_objects';
+import {
+  SavedObjectsExportResultDetails,
+  SavedObjectExportBaseOptions,
+  SavedObjectsExportByObjectOptions,
+  SavedObjectsExportByTypeOptions,
+} from './types';
+
+/**
+ * @public
+ */
+export type ISavedObjectsExporter = PublicMethodsOf<SavedObjectsExporter>;
+
+/**
+ * @public
+ */
+export class SavedObjectsExporter {
+  private savedObjectsClient: SavedObjectsClientContract;
+  private exportSizeLimit: number;
+
+  constructor({
+    savedObjectsClient,
+    exportSizeLimit,
+  }: {
+    savedObjectsClient: SavedObjectsClientContract;
+    exportSizeLimit: number;
+  }) {
+    this.savedObjectsClient = savedObjectsClient;
+    this.exportSizeLimit = exportSizeLimit;
+  }
+
+  /**
+   * Generates an export stream for given types.
+   *
+   * See the {@link SavedObjectsExportByTypeOptions | options} for more detailed information.
+   */
+  public async exportByTypes(options: SavedObjectsExportByTypeOptions) {
+    const objects = await this.fetchByTypes(options);
+    return this.processObjects(objects, options);
```

Comment:
suggestion: I'd separate method-specific options from base export options to prevent leakage to `processObjects`.
```typescript
public async exportByTypes(types: ExportByType, options: ExportOptions) {
```

Reply:
I would like to avoid splitting the parameters into two structs. I updated the calls to `processObjects` to explicitly only include the properties defined in `SavedObjectExportBaseOptions`